cmd/c2w-net/main.go

```diff
@@ -139,12 +176,108 @@ func main() {
 	if err != nil {
 		panic(err)
 	}
-	// We register our VM network as a qemu "-netdev socket".
 	if err := vn.AcceptQemu(context.TODO(), conn); err != nil {
 		panic(err)
 	}
 }
 
+func waitForContainerIP(vn *gvnvirtualnetwork.VirtualNetwork) string {
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+	timeout := time.After(10 * time.Second)
+
+	for {
+		select {
+		case <-timeout:
+			log.Printf("timeout waiting for DHCP lease, falling back to default IP %s\n", vmIP)
+			return vmIP
+		case <-ticker.C:
+			if ip, err := getContainerIP(vn); err == nil {
+				return ip
+			}
+		}
+	}
+}
+
+func getContainerIP(vn *gvnvirtualnetwork.VirtualNetwork) (string, error) {
+	req := httptest.NewRequest("GET", "/services/dhcp/leases", nil)
+	w := httptest.NewRecorder()
+	vn.ServicesMux().ServeHTTP(w, req)
+
+	if w.Code != http.StatusOK {
+		return "", fmt.Errorf("failed to query DHCP leases: status %d", w.Code)
+	}
+
+	var leases map[string]string
+	if err := json.NewDecoder(w.Body).Decode(&leases); err != nil {
+		return "", fmt.Errorf("failed to decode DHCP leases: %v", err)
+	}
+
+	for ip := range leases {
+		if ip != gatewayIP {
+			return ip, nil
+		}
+	}
+
+	return "", fmt.Errorf("no container leases found")
+}
+
+func startPortForwarder(localAddr, remoteAddr string, vn *gvnvirtualnetwork.VirtualNetwork) {
+	startPortForwarderWithContext(context.Background(), localAddr, remoteAddr, vn)
+}
+
+func startPortForwarderWithContext(ctx context.Context, localAddr, remoteAddr string, vn *gvnvirtualnetwork.VirtualNetwork) {
+	listener, err := net.Listen("tcp", localAddr)
+	if err != nil {
+		log.Printf("failed to listen on %s: %v\n", localAddr, err)
+		return
+	}
+	defer listener.Close()
+	log.Printf("forwarding %s -> %s\n", localAddr, remoteAddr)
+
+	go func() {
+		<-ctx.Done()
+		listener.Close()
+	}()
+
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				log.Printf("port forwarder %s stopped\n", localAddr)
+				return
+			default:
+				log.Printf("failed to accept connection on %s: %v\n", localAddr, err)
+				return
+			}
+		}
+		go handleConnection(conn, remoteAddr, vn)
+	}
+}
+
+func handleConnection(clientConn net.Conn, remoteAddr string, vn *gvnvirtualnetwork.VirtualNetwork) {
+	defer clientConn.Close()
+
+	targetConn, err := vn.DialContextTCP(context.Background(), remoteAddr)
+	if err != nil {
+		log.Printf("failed to dial %s: %v\n", remoteAddr, err)
+		return
+	}
+	defer targetConn.Close()
+
+	done := make(chan struct{}, 2)
+	go func() {
+		io.Copy(targetConn, clientConn)
+		done <- struct{}{}
+	}()
+	go func() {
+		io.Copy(clientConn, targetConn)
+		done <- struct{}{}
+	}()
+	<-done
```

Comment:
nit: I guess either of the coroutines can leak here with waiting for `done <- struct{}{}` forever without the receiver.  To avoid this issue, `sync.WaitGroup` should be used here.